src/azul/indexer/log_forwarding_service.py

```diff
@@ -0,0 +1,110 @@
+import csv
+import gzip
+from typing import (
+    Iterator,
+)
+
+from botocore.response import (
+    StreamingBody,
+)
+from furl import (
+    urlsplit,
+)
+
+from azul import (
+    require,
+)
+from azul.deployment import (
+    aws,
+)
+from azul.types import (
+    MutableJSON,
+)
+
+
+class LogForwardingService:
+    """
+    Parse logs from an Application Load Balancer.
+    Adapted from https://github.com/rupertbg/aws-load-balancer-logs-to-cloudwatch
+    """
+
+    def read_logs(self, bucket: str, key: str) -> Iterator[MutableJSON]:
+        response = aws.s3.get_object(Bucket=bucket, Key=key)
+        body: StreamingBody = response['Body']
+        with gzip.open(body, mode='rt') as f:
+            # CSV format escapes the quotechar by repeating it. This cannot
+            # occur in the logs because quotations marks occurring within the
+            # field values are represented as '\x22'.
+            for row in csv.reader(f, delimiter=' ', quotechar='"'):
+                # When new fields are introduced, they are added at the end of
+                # the log entry, so observing more fields than expected does not
+                # indicate a problem.
+                require(len(row) >= len(self.fields), 'Missing expected fields')
+                fields = {
+                    k: self._restore_escaped_chars(v)
+                    for k, v in zip(self.fields, row)
+                }
+                self._expand_fields(fields)
+                yield fields
+
+    def _expand_fields(self, message: MutableJSON) -> None:
+        message.update(self._parse_request(message['request']))
+        for field_name in ['target:port', 'client:port']:
+            message.update(self._parse_port(field_name, message[field_name]))
+
+    def _restore_escaped_chars(self, field: str) -> str:
+        # https://stackoverflow.com/a/58829514/1530508
+        return field.encode().decode('unicode_escape').encode('latin1').decode()
```

Comment:
```suggestion
        return field.encode('ascii').decode('unicode-escape')
```

or

```suggestion
        return field.encode('latin1').decode('unicode-escape')
```

would be sufficient for our case.

Reply:
These approaches do not correctly decode non-ASCII characters. 

I tested with `È` (0xc8, in latin-1), `ā` (0x101, Latin extended A), `ȁ`, (0x201, Latin extended B), `Ё` (0x401, Cyrillic), and 😀 (0x1f600, Emoticons). These were encoded in the log files as `\xC3\x88`, `\xC4\x81`, `xC8\x81`, and `\xf0\x9f\x98\x80` respectively.

```
>>> escaped = [r'\xC3\x88', r'\xC4\x81', r'\xC8\x81', r'\xf0\x9f\x98\x80']
>>> [s.encode().decode('unicode_escape').encode('latin1').decode() for s in escaped]
['È', 'ā', 'ȁ', '😀']
>>> [s.encode('ascii').decode('unicode_escape') for s in escaped]
['Ã\x88', 'Ä\x81', 'È\x81', 'ð\x9f\x98\x80']
>>> [s.encode('latin1').decode('unicode_escape') for s in ss]
['Ã\x88', 'Ä\x81', 'È\x81', 'ð\x9f\x98\x80']
```

Reply:
To reproduce:
```
for x in 200 257 513 1025; do 
    curl -v 'https://gitlab.dev.singlecell.gi.ucsc.edu/EXPERIMENT?foo='$(python3 -c "print(chr($x))"); 
done
...
AWS_PROFILE=platform-hca-dev aws s3 cp <s3 uri of log file> - | gunzip | grep EXPERIMENT | cut -d' ' -f14
```

Reply:
See also: https://github.com/python/cpython/issues/74773#issuecomment-1093750401

Reply:
The problem with this approach is that the string returned by `decode(`unicode_escape`)` is encoded using `latin1` instead of `utf8`. Every non-ASCII character is escaped in the log file using its `utf8` encoding consisting of at least 2 bytes (based on experiments), so even `latin1`-compatible characters aren't correctly decoded in a single step. See https://docs.python.org/3/library/codecs.html#text-encodings